### PR TITLE
feat: add Boyer-Moore algorithm

### PR DESCRIPTION
**Pull request by Soptq**
This PR adds Boyer-Moore algorithm.

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

- searching pattern ByteArray in a text ByteArray using Boyer-Moore algorithm.
- all tests are passed.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/searching/src/bm_search.cairo [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// The Boyer-Moore string search algorithm
+use dict::Felt252DictTrait;
+
+// Find `pattern` in `text` and return the index of every match.
```
Comment (LucasLvy) [inline src/searching/src/bm_search.cairo:4]:
```suggestion
/// Find `pattern` in `text` and return the index of every match.
```

File: src/searching/src/bm_search.cairo [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+/// # Returns
+/// * `Array<usize>` - The index of every match.
+fn bm_search(text: @ByteArray, pattern: @ByteArray) -> Array<usize> {
+    let mut positions: Array<usize> = array![];
+    let n = text.len();
+    let m = pattern.len();
+    if n == 0 || m == 0 || m > n {
+        return positions;
+    }
+
```
Comment (LucasLvy) [inline src/searching/src/bm_search.cairo:16]: add comments and add meaningful variable names please

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: src/searching/src/bm_search.cairo (hunk 1 above)
Comment (Soptq) [reply src/searching/src/bm_search.cairo]: Fixed ;)

File: src/searching/src/bm_search.cairo (hunk 2 above)
Comment (Soptq) [reply src/searching/src/bm_search.cairo]: Fixed ;)

**Pull request merged**